### ISSUE TITLE
[frontend] Expression evaluation fails for `await` statements with a compilation error

### ISSUE THREAD

**Issue by annagrin** [dart-lang/sdk#45454]
Currently expression compilation does not support `await` expressions  - looks like the synthetic debug function is just always created as a sync function. Are there any known limitations to supporting `await`? The question comes up in conversations with customers, we will know more about priority soon after user feedback collection.

Repro:

- debug flutter hello_world example (for any platform, my output below is for web)
- set a breakpoint at `runApp`
- on breakpoint, type `await (() async => "foo")()`

Result:

CompilationError: Method not found: 'await'.
await (() async => "foo")()
^^^^^

Expected:
'foo'

**Comment by mraleph** [comment]
Async machinery (including propagation of `Future` completions) is handled by the event loop so you can't really `await` something while stopped on a breakpoint. 

This is simply going to "dead lock". You need to run the event loop: this either means running a nested event loop or actually continuing from the breakpoint. Nested event loop is a can of worms (and probably impossible in the browser anyway), which leaves you with the only option of continuing from the breakpoint. 

But in that case it's not really "evaluation of an expression on the breakpoint". 

See @lrhn answer [here](https://github.com/dart-lang/sdk/issues/44999#issuecomment-778649786).

**Comment by annagrin** [comment]
@mraleph I suspected something along those lines, good to have a confirmation!

**Comment by mraleph** [comment]
I think it is safe to close this as working as intended. We could consider doing a better error message though.

**Comment by mraleph** [comment]
Actually I thought about this a bit more and there _might_ be a way around this, though it is going to be a pretty gnarly thing to implement. I am going to add my thinking to #44999

**Comment by mraleph** [comment]
Duplicate of #44999

**Comment by annagrin** [comment]
> Duplicate of #44999

@mraleph The bug you mentioned seems to only talk about VM side of things - while the current one mentions the frontend - incremental compiler cannot currently compile `await` expressions. Should both be open?

**Comment by mraleph** [comment]
Yeah, we can open both and use one as an umbrella. We also need to understand if this is implementable in the browser.
